torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -494,6 +496,256 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+    if (dom[i]->isBroadcast()) {
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+    } else {
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
+    }
+  }
+
+  // This domain will be the consumer, so create the producer
+  TensorView* producer =
+      new TensorView(new TensorDomain(out_domain), this->getDataType().value());
+
+  // Set domain of consumer
+  TensorView* consumer = this;
+
+  // Setup dependency chain, inserting producer - Cache_Read - before this op.
+  // Before: Prev Op -> This Op
+  // After: Prev Op -> New Cache_Read Op -> This Op
+
+  // Expr* producer_origin =
+  if (origin_expr != nullptr)
+    this->createExprConsumer(origin_expr, producer);
+
+  // Expr* producer_uses =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // ComputeAt - Before
+  if(this->hasComputeAt())
+    producer->computeAt(consumer, getThisComputeAtAxis());
+
+  return producer;
+}
+
+// Create a TensorView after the original tensor. A common use case is to write
+// results into shared memory or local registers before moving to global memory.
+// Analogous to TVM Cache_Read
+TensorView* TensorView::cache_after() {
+  FusionGuard fg(this->fusion());
+
+  // Get all the uses for this Tensorview
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
+  TORCH_CHECK(
+      uses_expr.size() > 0,
+      "Error adding cache_after ",
+      this,
+      " its #uses is zero and we restrict using cache_after on an output.");
+
+  // Create Producer Domain
+  auto dom = domain()->noReductions();
+  std::vector<IterDomain*> out_domain(dom.size(), nullptr);
+  for (size_t i = 0; i < dom.size(); ++i) {
+    if (dom[i]->isBroadcast()) {
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+    } else {
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
+    }
+  }
+
+  // This domain will be the producer, so create the consumer
+  TensorView* consumer =
+      new TensorView(new TensorDomain(out_domain), this->getDataType().value());
+
+  // Set domain of producer - No Change
+  TensorView* producer = this;
+
+  // Setup dependency chain, inserting consumer - Cache_Read - after this op.
+  // Before: This Op -> Next Op
+  // After: This Op -> New Cache_Write Op -> Next Op
+
+  // Expr* consumer_origin =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // Expr* consumer_uses =
+  for (auto expr : uses_expr) {
+    createExprProducer(expr, consumer);
+    //this->fusion()->removeExpr(expr);
+  }
+
+  // ComputeAt - After
+  if(this->hasComputeAt())
+    producer->computeAt(consumer, getThisComputeAtAxis());
+
+  return consumer;
+}
+
+// Create New Expr given producer - [an input for the expression]
+void TensorView::createExprProducer(Expr* expr, TensorView* producer) {
+  switch (*(expr->getExprType())) {
+    case ExprType::Split:
```

Comment:
Use one of the dispatch mechanisms in dispatch.h you can find many examples of usage in the code base.

Reply:
Just a warning that I didn't carefully look at below.